### ISSUE TITLE
Deeplinking into client apps

### ISSUE THREAD

**Issue by saket** [mastodon/mastodon#20205]
### Pitch

One of the biggest inconveniences of using mastodon on mobile is that client apps can't support deeplinks. When a mastodon url is clicked from another app, it opens in the web browser instead of an installed client app making it difficult to, say, follow a user on other servers. 

Although it is possible for client apps to declare a list of all known mastodon hosts at compile time, it'll be a very bad idea considering there can be infinite mastodon servers.

Have there been any discussions around this?

### Motivation

Deeplinking will allow users to view mastodon urls in their client apps instead of landing in a web browser, where they may not be signed in.

**Comment by Gargron** [comment]
Deep-linking in iOS and Android relies on a hardcoded list of domains, which is incompatible with the dynamic and decentralized nature of the fediverse. As much as we'd like to, we can't do this unless Apple and Google do something to support our use-case.

**Comment by saket** [comment]
I've filed a feature request for Android: https://issuetracker.google.com/issues/258270055

**Comment by ashdavies** [comment]
It might be possible for a client app to rewrite a url with a `mastodon://` prefix to then handle links in app, but this wouldn't handle links outside of said app

**Comment by errotu** [comment]
> A workaround could be to implement a mastodon url scheme which the app hooks into. When opening an URL, users would still be directed to the browser but at this point the user could be given the option to open the url in the app.

I would like to second Laura's idea, if I understood her correctly. This could be reached in a similar manner like the Signal app does [its set-up for its proxies](https://signal.org/blog/run-a-proxy/). signal.tube is mere referrer to the actual proxy address. All shares could be channeled through such a service.

For example, this post: https://mastodon.social/@Gargron/109300967725833789
could be reached via https://mastodon.link/#mastodon.social/@Gargron/109300967725833789

The places were links to a Toot exist (the date and the copy-link-option in the three-dots-menu) should then link to the referrer, not the actual Toot.

The only hard-coded domain within the app would then be mastodon.link. The biggest caveat I see would be the introduction of a single point of failure when it comes to links to Mastodon posts. This could be mitigated, however, by making the redirection via mastodon.link only an option for server administrators.

**Comment by LauraWebdev** [comment]
> > A workaround could be to implement a mastodon url scheme which the app hooks into. When opening an URL, users would still be directed to the browser but at this point the user could be given the option to open the url in the app.
> 
> I would like to second Laura's idea, if I understood her correctly. This could be reached in a similar manner like the Signal app does [its set-up for its proxies](https://signal.org/blog/run-a-proxy/). signal.tube is mere referrer to the actual proxy address. All shares could be channeled through such a service.
> 
> For example, this post: https://mastodon.social/@Gargron/109300967725833789 could be reached via https://mastodon.link/#mastodon.social/@Gargron/109300967725833789
> 
> The places were links to a Toot exist (the date and the copy-link-option in the three-dots-menu) should then link to the referrer, not the actual Toot.
> 
> The only hard-coded domain within the app would then be mastodon.link. The biggest caveat I see would be the introduction of a single point of failure when it comes to links to Mastodon posts. This could be mitigated, however, by making the redirection via mastodon.link only an option for server administrators.

I'm guessing this isn't in line with the whole decentralized approach of mastodon. However I did not see a problem using HTTP schemes (at least on android) without a host. A custom scheme could include the host like this: mastodon://tech.lgbt

**Comment by errotu** [comment]
Hm, on the other hand, such links could not be opened by a regular browser, if someone has no Mastodon app installed. 

A regular use-case for me is forwarding the link to a Toot via another messenger (e.g. Signal). If the person uses Mastodon, it should be able to open the link within their app, if not, the Toot should be opened by the browser.

**Comment by LauraWebdev** [comment]
> Hm, on the other hand, such links could not be opened by a regular browser, if someone has no Mastodon app installed.
> 
> A regular use-case for me is forwarding the link to a Toot via another messenger (e.g. Signal). If the person uses Mastodon, it should be able to open the link within their app, if not, the Toot should be opened by the browser.

The idea was to add a banner to mastodon instances when entering a post on a smartphone, giving the user the ability to open the link in the app (Similarly how some forums software do it).

**Comment by errotu** [comment]
Ah! Makes sense and would be a good approach, I'd say. Thank you for the explanation!

**Comment by cguess** [comment]
Has there been any movement on this? I'm a Ruby dev of 17 years and an iOS Dev of 11 (plus Android work), so I'm considering working on this but would like the blessing of the community before I spend the time and effort.

**Comment by rfulwell** [comment]
> Has there been any movement on this? I'm a Ruby dev of 17 years and an iOS Dev of 11 (plus Android work), so I'm considering working on this but would like the blessing of the community before I spend the time and effort.

Yes, please! With the latest influx of users, Mastodon needs this a month ago! 😅

**Comment by joshbalfour** [comment]
> Hm, on the other hand, such links could not be opened by a regular browser, if someone has no Mastodon app installed. 
> 
> 
> 
> A regular use-case for me is forwarding the link to a Toot via another messenger (e.g. Signal). If the person uses Mastodon, it should be able to open the link within their app, if not, the Toot should be opened by the browser.

It's possible for PWAs to register to open protocol links e.g. gmail.com asks to handle mailto urls

https://developer.mozilla.org/en-US/docs/Web/Manifest/protocol_handlers

I think this would be nicely in line with decentralisation, and allow users to choose the client or instance they want to use.

This would be an awesome fix for the user following flow UX (that's what brought me to this issue in the first place).

I'm happy to PR the web client side of this if the concepts likely to be signed off.

**Comment by vinaygopinath** [comment]
Trying to revive this thread:

Do I understand correctly that the consensus in this thread is to set up a website like mastodon.link that gives users the option to either continue in the web browser or open the link in any installed client app that has registered itself to handle mastodon.link?

![open-mastodon-link drawio](https://github.com/mastodon/mastodon/assets/324200/35a2126a-5748-4df5-be1b-b8cf4f40ab06)

@cguess Thanks for offering to take on this implementation! Is this something you're still looking into?

**Comment by strk** [comment]
> A workaround could be to implement a mastodon url scheme which the app hooks into. When opening an URL, users would still be directed to the browser but at this point the user could be given the option to open the url in the app.

There is a proposal for an web+ap:// url scheme here: https://fedilinks.org/
